gnomad_qc/v4/analyses/grpmax_comps.py

```diff
@@ -0,0 +1,294 @@
+    :param version: gnomAD version
+    :return: Table with European AF annotation
+    """
+    if version == "v4":
+        eur_indices = hl.literal(eur_grps).map(lambda g: ht.freq_index_dict[g + "_adj"])
+    else:
+        eur_indices = hl.literal(eur_grps).map(
+            lambda g: ht.freq_index_dict["gnomad_" + g]
+        )
+    ht = ht.annotate(
+        eur_AC=hl.sum(
+            eur_indices.map(
+                lambda f: hl.if_else(hl.is_defined(ht.freq[f].AC), ht.freq[f].AC, 0)
+            )
+        ),
+        eur_AN=hl.sum(
+            eur_indices.map(
+                lambda f: hl.if_else(hl.is_defined(ht.freq[f].AN), ht.freq[f].AN, 0)
+            )
+        ),
+    )
+    ht = ht.annotate(eur_AF=hl.if_else(ht.eur_AN > 0, ht.eur_AC / ht.eur_AN, 0.0))
+    return ht.drop("eur_AC", "eur_AN")
+
+
+def filter_to_threshold(
+    ht: hl.Table,
+    af_threshold: float = 0.001,
+    version: str = "v4",
+):
+    """
+    Filter to variants where eur AF is < threshold while grpmax > threshold.
+
+    :param ht: gnomAD release Table
+    :param af_threshold: AF dividing threshold
+    :param version: gnomAD version
+    :return: Table filtered to variants meething threshold specifications
+    """
+    if version == "v4":
+        grpmax_expr = ht.grpmax.gnomad
+    else:
+        grpmax_expr = ht.popmax[0]
+    ht = ht.filter((grpmax_expr.AF > af_threshold) & (ht.eur_AF < af_threshold))
+    return ht
+
+
+def version_stats(
+    ht: hl.Table,
+    version: str = "v4",
+    can_only: bool = False,
+    drop_hard_filtered_variants: bool = False,
+    all_variants: bool = False,
+) -> Dict[str, Dict[str, Dict[str, int]]]:
+    """
+    Calculate grpmax stats for a given gnomAD version.
+
+    :param ht: gnomAD release Table
+    :param version: gnomAD version
+    :param can_only: Only consider MANE Select and canonical transcripts.
+    :param drop_hard_filtered_variants: Remove only hard filtered variants but keep all other variants regardless of variant QC status
+    :param all_variants: Keep all variants regardless of variant QC status
+    :return: Dictionary of grpmax stats
+    """
+    logger.info("Calculating grpmax stats for %s", version)
+    # Group results into nfe and all eur grps
+    results_by_eur_grping = {}
+
+    logger.info(
+        "Total number of variants in %s before filtering: %s", version, ht.count()
+    )
+    t_variants = ht.count()
+    if can_only:
+        logger.info("Filtering to only MANE Select and canonical transcripts")
+        ht = filter_vep_to_canonical_transcripts(ht, filter_empty_csq=True)
+        # All MANE select transcripts in v4 are also the canonical transcript
+        logger.info("Filtering to keep only non-synonymous variants...")
+        ht = ht.filter(
+            hl.any(
+                ht.vep.transcript_consequences.consequence_terms.map(
+                    lambda x: hl.literal(NS_CONSEQ_TERMS).contains(x)
```

Comment:
re-reading this and realizing maybe it's better to use [`filter_vep_transcript_csqs`](https://github.com/broadinstitute/gnomad_methods/blob/main/gnomad/utils/vep.py#L717) to simultaneously filter to canonical transcripts and on synonymous variants

Reply:
I restructured this and actually tried using this function but hit a bug: https://github.com/broadinstitute/gnomad_methods/issues/707

Reply:
ooh. I guess this function needs to get run first? https://github.com/broadinstitute/gnomad_methods/blob/019865838f993841a540e0b29d8d2f3b1333b1b8/gnomad/utils/vep.py#L273